praxis/layers/fp8_ops.py

```diff
@@ -0,0 +1,186 @@
+  return 1.0 / sf
+
+def compute_scale_and_amax_history(x, q_dtype, scale, amax_history):
+  dtype_max = get_fp8_max(q_dtype, jnp.float32)
+
+  amax_update = jnp.max(jnp.abs(x)).astype(scale.dtype)
+  new_amax_history = \
+      jnp.roll(amax_history, shift=-1, axis=0).at[0].set(amax_update)
+
+  amax_from_history = jnp.max(new_amax_history, axis=0)
+  new_scale = compute_scale(amax_from_history, scale, dtype_max)
+  return new_scale, new_amax_history
+
+def qdq_and_return(x, q_dtype, scale, amax_history, compute_dtype):
+  qx = quantize_dequantize(x, q_dtype, scale, compute_dtype)
+  new_scale, new_amax_history = compute_scale_and_amax_history(
+      x, q_dtype, scale, amax_history)
+  return qx, new_scale, new_amax_history
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def in_qdq(compute_dtype, inp, scale, amax_history):
+  qin, _, _ = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin
+
+def in_qdq_fwd(compute_dtype, inp, scale, amax_history):
+  qin, new_scale, new_amax_history = qdq_and_return(
+      inp, jnp.float8_e4m3fn, scale, amax_history, compute_dtype)
+  return qin, (new_scale, new_amax_history)
+
+def in_qdq_bwd(compute_dtype, res, g):
+  new_scale, new_amax_history = res
+  q_g = g
+  return q_g, new_scale, new_amax_history
+
+in_qdq.defvjp(in_qdq_fwd, in_qdq_bwd)
+
+
+@partial(custom_vjp, nondiff_argnums=(0,))
+def out_qdq(compute_dtype, out, scale, amax_history):
+  return out
+
+def out_qdq_fwd(compute_dtype, out, scale, amax_history):
+  return out, (scale, amax_history)
+
+def out_qdq_bwd(compute_dtype, res, g):
+  scale, amax_history = res
+  q_g, new_scale, new_amax_history = qdq_and_return(
+      g, jnp.float8_e5m2, scale, amax_history, compute_dtype)
+  return q_g, new_scale, new_amax_history
+  
+out_qdq.defvjp(out_qdq_fwd, out_qdq_bwd)
+
+
+def fp8_einsum(eqn, x, k, compute_dtype, x_scale, x_amax_history, k_scale,
+               k_amax_history, dy_scale, dy_amax_history):
+  """Perform any einsum formula.  """
+
+  x_qdq = in_qdq(compute_dtype, x, x_scale, x_amax_history)
+
+  k_qdq = in_qdq(compute_dtype, k, k_scale, k_amax_history)
+
+  y_qdq = jnp.einsum(eqn, x_qdq, k_qdq)
+
+  y = out_qdq(compute_dtype, y_qdq, dy_scale, dy_amax_history)
+
+  return y
+
+
+class Fp8EinsumOp(base_layer.BaseLayer):
+  """Wrapper around jnp.einsum used in standard Pax layers."""
+  amax_history_length: int = 1024
+
+  def setup(self) -> None:
+    scale_args = {
+        'shape': [1],
+        'init': WeightInit.Constant(1.0),
+        'dtype': jnp.float32,
+        'mesh_shape': self.mesh_shape,
+        'tensor_split_dims_mapping': None,
+        'collections': ['fp8_params'],
+    }
+    amax_history_args = {
+        'shape': [self.amax_history_length],
+        'init': WeightInit.Constant(0.0),
+        'dtype': jnp.float32,
+        'mesh_shape': self.mesh_shape,
+        'tensor_split_dims_mapping': None,
+        'collections': ['fp8_params'],
+    }
+    self.create_variable(
```

Comment:
Emm, after I put them into the non-trainable, it seems the parameters won't be put to the self.theta. So I got:
```
Traceback (most recent call last):
  File "/home/workspace/repo_zoo/praxis/praxis/layers/linears_test.py", line 515, in test_fp8_einsum_injection
    output2a, output2b = run(einsum_tpl, expected_shapes_new)
  File "/home/workspace/repo_zoo/praxis/praxis/layers/linears_test.py", line 475, in run
    initial_vars = ffn.init(
  File "/opt/flax/flax/linen/module.py", line 602, in wrapped_module_method
    return self._call_wrapped_method(fun, args, kwargs)
  File "/opt/flax/flax/linen/module.py", line 1119, in _call_wrapped_method
    y = run_fun(self, *args, **kwargs)
  File "/home/workspace/repo_zoo/praxis/praxis/base_layer.py", line 1804, in init
    variables = super().init(
  File "/home/workspace/repo_zoo/praxis/praxis/layers/linears.py", line 99, in __call__
    out = self.einsum('...y,yz->...z', inputs, self.theta.w)
  File "/home/workspace/repo_zoo/praxis/praxis/layers/fp8_ops.py", line 186, in __call__
    out = fp8_einsum(equation, x, k, comp_dtype, theta.input_scale,
  File "/home/workspace/repo_zoo/praxis/praxis/base_layer.py", line 1236, in __getattr__
    raise ValueError(f'Module {self.module} has no theta.{k} defined.')
ValueError: Module Fp8EinsumOp(
    # attributes
    dtype = float32
    fprop_dtype = float32
    params_init = WeightInit(method='xavier', scale=1.000001)
    skip_lp_regularization = None
    ici_mesh_shape = None
    dcn_mesh_shape = None
    contiguous_submeshes = None
    mesh_axis_names = None
    shared_weight_layer_id = None
    weight_split_dims_mapping = <PaxConfig[BaseLayer.WeightSharding()]>
    activation_split_dims_mapping = <PaxConfig[BaseLayer.ActivationSharding()]>
    amax_history_length = 1024
) has no theta.input_scale defined.
```

Reply:
No worry about the above. I figured out that I can use `self.get_var()` for that.

I have played with the `non_trainable` idea, yes, it can have the grads. But the tricky part is that still many places need to be changed accordingly because the places assume that non-trainable == no grads, for example, the logics in get_excluded_var_mask_for_grad which would exclude grad for all non trainable. Even if I fix this logic, I still cannot make it run. Then, I feel I should still go back to the old way since no matter which we still need the `fp8_params` flag. But I simplify the code a bit to make the fp8 logic stand out/more clear.